### ISSUE TITLE
Migrate rendering plugins to `SurfaceProducer` after stable release

### ISSUE THREAD

**Issue by matanlurey** [flutter/flutter#139702]
**tl;dr**: 

- It is both non-trivial to support the Android `SurfaceTexture` API in Impeller (on Android with Vulkan)
- `SurfaceTexture` is not a desirable integration point (based on talking to the Android team directly)
- Usage is quite low, and there is an easy/mechanical path to migrate it to something better

## Background

The Flutter engine on Android supports two external rendering sources, `SurfaceTexture`[^1][^2][^3] (an OpenGLES texture) and `ImageReader`[^4][^5][^6] (a GPU-ready buffer). [`Image.getHardwareBuffer`](https://developer.android.com/reference/android/media/Image#getHardwareBuffer), which is used by our `ImageReader`-based code path, is only supported on newer Android API versions (API level 28).

In Impeller, for the first time, we'll support [Vulkan](https://developer.android.com/ndk/guides/graphics), [starting at Android API version 29](https://github.com/flutter/engine/blob/6f5b6347d472dc45197305449c9cb06259c210ff/shell/platform/android/platform_view_android.cc#L82-L101), over OpenGLES (all Flutter Android apps currently use Skia, which in turn uses OpenGLES).  Unfortunately, it's non-trivial to use an OpenGLES texture (`SurfaceTexture`) in Vulkan, and after some internal discussion with the Android team, we've come to the conclusion that `SurfaceTexture` is not the right integration point in general.

I did a quick scan of _usage_ of `(create|register)SurfaceTexture`, both internally and externally:

- Internal numbers are small (single digits)
- Within `org: flutter`, we have [about 17 usages](https://github.com/search?q=org%3Aflutter+%2F%28create%7Cregister%29SurfaceTexture%2F&type=code)
- Github-wide, I see [<400 usages](https://github.com/search?q=-org%3Aflutter+%2F%28create%7Cregister%29SurfaceTexture%2F+flutter++NOT+is%3Aarchived+NOT+is%3Afork+language%3AJava+OR+language%3AKotlin&type=code), many of which are ancient projects, forks, or otherwise not critical ecosystem pieces

[^1]: [`android.graphics.SurfaceTexture`](https://developer.android.com/reference/android/graphics/SurfaceTexture?hl=en)
[^2]: [`surface_texture_external_texture.h`](https://github.com/flutter/engine/blob/main/shell/platform/android/surface_texture_external_texture.h)
[^3]: [`platform_view_android.cc`](https://github.com/flutter/engine/blob/fe253b0460e6e5230e5a9e681d5166f7c13b93f8/shell/platform/android/platform_view_android.cc#L331-L333)
[^4]: [`android.media.ImageReader`](https://developer.android.com/reference/android/media/ImageReader)
[^5]: [`image_external_texture.h`](https://github.com/flutter/engine/blob/main/shell/platform/android/image_external_texture.h)
[^6]: [`platform_view_android.cc`](https://github.com/flutter/engine/blob/fe253b0460e6e5230e5a9e681d5166f7c13b93f8/shell/platform/android/platform_view_android.cc#L352-L354)

## Proposal

We'll add a new, parallel API/extension point, i.e. `(create|register)Surface`[^7]. Similar to our platform view mechanics, we'll _transparently_ either use `SurfaceTexture` (on older devices) or `ImageReader` (on newer devices), and end-users (users that consume plugins that use `(create|register)Surface` will not have to do or configure anything.

[^7]: [`android.view.Surface`](https://developer.android.com/reference/android/view/Surface)

## Timeline

- [x] 1. Add a new API, `(create|register)Surface`, which delegates (internally) to `SurfaceTexture` or `ImageReader`
- [x] 2. Migrate internal plugins
- [x] 3. Migrate `org: flutter` plugins and examples
- [x] 4. Add user-facing documentation on flutter.dev about the recommended approach
- [x] 5. Deprecate `(create|register)SurfaceTexture`. In Impeller, issue a warning at runtime if it is used (Proposal: https://github.com/flutter/flutter/issues/148557)
- [x] 6. In Impeller, issue an error if it is used (Done: https://github.com/flutter/engine/pull/52892)
- [ ] 7. By the time Skia-Ganesh backend is removed, remove `(create|register)SurfaceTexture` entirely

## Testing Strategy

Our ability to run comprehensive Android device tests in the engine repository is very limited (as of 2023-12-06):

- Android engine unit tests are mock-heavy
- Separation of framework and engine means to properly write an integration test you need code on "both" sides

@johnmccutchan coincidentally has a napkin-math proposal to improve the state of Android platform testing, which is critical for stability of our Android platform views and plugins. As necessary I plan to pair with him to iron out that proposal, land it, and then consume whatever infrastructure we built out/refactor out to _also_ test our `(create|register)Surface` API.

Some scenarios will include:

- An older Android {device, emulator} that will implicitly use surface textures
- A newer Android {device, emulator} that will implicitly use hardware buffers (via `ImageReader`)
- Try dynamically resizing the texture at runtime
- Try "animated" textures (i.e. each frame is a different image)
- Try tearing down and re-setting up views that use external textures

---

/cc @dnfield @jonahwilliams @gaaclarke @johnmccutchan @reidbaker

**Comment by jonahwilliams** [comment]
The timeframe between 5 and 6/7 is likely the better part of a year. We do need to introduce these APIs soon, but migrating the ecosystem can wait.

> Why would waiting until after the next stable to migrate our plugins be an issue?

Its only an issue in the very literal sense that it is something to do that is tracked in the issue tracker, I'm not implying there would be a problem with it. Only that there would be a problem if we failed to consider it.

**Comment by matanlurey** [comment]
Thanks! There is no need for an aggressive timeline to migrate; the only caveat is the plugins that use SurfaceTexture today will *not* work on Impeller until they are migrated.

Based on that, it sounds like we would need to migrate the plugins at least 2 releases before Impeller reaches stable?

**Comment by matanlurey** [comment]
> Is the migration here basically a one-line change, or are there complex changes needed as side-effects of changing the surface type?

In theory it's not a _complex_ change, but for example, [looking at `camerax`](https://github.com/flutter/packages/blob/15584a394d36070b684b58bb65b37cd86aa1f1ea/packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java#L65), it will require more than a 1-line change.

I'll prototype a bit with @johnmccutchan before I say anything definitive though.

**Comment by gaaclarke** [comment]
It's only trivial if the consumer of the buffer takes into the account that they may be getting a vulkan buffer, like `androidx.camera.core.Preview` in camerax, right?  If they do anything with the buffer beyond pass it to something that already accounts for it, they will be broken.

I can imagine things like OpenCV plugins stopping to work.  We want those plugins to be blocked at compilation time somehow.  Removing the old API would do that.  What happens when someone makes a plugin that only supports vulkan or opengles?

Can we enforce a fallback to impeller opengles if you have a plugin that only supports it?

Instead of creating an abstraction over ImageReader and SurfaceTexture, can we just make plugins implement both APIs?  Which one is used will be determined at runtime.  The SurfaceTexture, if it ever looks into the opengl will just be a subset of the implementation of the ImageReader, right?  Eventually SurfaceTexture could be removed once the android sdk for Flutter gets bumped.

**Comment by matanlurey** [comment]
@gaaclarke:

> Can we enforce a fallback to impeller opengles if you have a plugin that only supports it

I think we would know "too late". The last time I talked to @jonahwilliams at least, we didn't want to _allow_ users to specify they wanted the OpenGLES background specifically. We of course could make a different decision, but that personally makes me nervous.

> Instead of creating an abstraction over ImageReader and SurfaceTexture, can we just make plugins implement both APIs?

That was the other option! We figured that (a) this would require more work for end-users and (b) it would be different from the expectation of platform views, where we handle this for users. Definitely could be revisited if our original plan is flawed :)

**Comment by johnmccutchan** [comment]
This bug is about the abstraction that the engine exposes so that native code can populate a texture that is composited in their Flutter app. 

The correct abstraction for the engine to offer is basically just a `Surface` that the application can use to render. This ensures that we hide the entire consumer part of the system and avoids issues in the future when say `ImageReader` is deprecated for something else. A Surface be rendered to from many APIs and we don't need to care how the plugins render their content, just that in the end they write it to the Surface we vend from the engine.

**Comment by stuartmorgan** [comment]
> I think we would know "too late".

When would we need to know? We know at compile time of the app what all the plugins are and what their lower Flutter constraint is, for instance, so if build time is early enough to enable a fallback mechanism we could gate that on min Flutter SDK version, much like Dart gates features on min Dart SDK version.

**Comment by stuartmorgan** [comment]
Yes; if we have a way to do fallback to not break existing plugins, but can't toggle it at runtime, then we could consider gating it on SDK constraint at build time. This allows deprecating the old path for new plugins, without breaking old plugins.

**Comment by matanlurey** [comment]
Quick 🧠 dump before the weekend (I might make more progress, just in case though):

@johnmccutchan landed the concept of a `SurfaceProducer`; https://github.com/flutter/engine/pull/48803:

```java
// shell/platform/android/io/flutter/embedding/engine/renderer/FlutterRenderer.java
public class FlutterRenderer implements TextureRegistry {
  /**
   * Creates and returns a new external texture {@link SurfaceProducer} managed by the Flutter
   * engine that is also made available to Flutter code.
   */
  @Override
  public SurfaceProducer createSurfaceProducer() { /* ... */ }
}
```

... and landed initial support for one backed by an `ImageReader`:

```java
final class ImageReaderSurfaceProducer implements TextureRegistry.SurfaceProducer { /* ... */ }
```

... tested in `FlutterRendererTest`.

We use JNI to access this class in the engine:

```cc
// shell/platform/android/platform_view_android_jni_impl.cc
PlatformViewAndroidJNIImpl::ImageProducerTextureEntryAcquireLatestImage(/*...*/) { /* ... */}
```

---

My next step is implementing the `SurfaceTextureSurfaecProvider` (and friends).

@jonahwilliams thinks I'll be able to verify this using the [`video_player/example`](https://github.com/flutter/packages/tree/main/packages/video_player/video_player/example).

**Comment by matanlurey** [comment]
Blocked by https://github.com/flutter/flutter/issues/141636.

**Comment by matanlurey** [comment]
Update: @camsim99 landed https://github.com/flutter/packages/pull/6856 (`camerax`) and https://github.com/flutter/packages/pull/6989 is in the merge queue (`video_player`), which will, once merged, complete this issue out.
